index.js

```diff
@@ -404,84 +404,85 @@ FeatureService.prototype._requestFeatures = function (task, cb) {
+        self.catchErrors(task, err, uri, cb)
       })
 
       response.on('end', function () {
-        // TODO: move this into a function call decode
-        try {
-          var json
-          var buffer = Buffer.concat(data)
-          var encoding = response.headers['content-encoding']
-          // TODO all this shit is ugly -- make it less shitty (less try/catch)
-          if (encoding === 'gzip') {
-            zlib.gunzip(buffer, function (e, result) {
-              try {
-                json = JSON.parse(result.toString().replace(/NaN/g, 'null'))
-                cb(null, json)
-              } catch (e) {
-                catchErrors(task, e, uri, cb)
-              }
-            })
-          } else if (encoding === 'deflate') {
-            try {
-              json = JSON.parse(zlib.inflateSync(buffer).toString())
-              cb(null, json)
-            } catch (e) {
-              catchErrors(task, e, uri, cb)
-            }
-          } else {
-            json = JSON.parse(buffer.toString().replace(/NaN/g, 'null'))
-            cb(null, json)
-          }
-        } catch(e) {
-          catchErrors(task, e, uri, cb)
-        }
+        self.decode(response, data, function (err, json) {
+          if (err) return self.catchErrors(task, err, uri, cb)
+          cb(null, json)
+        })
       })
+
     })
 
     req.setTimeout(self.timeOut, function () {
       // kill it immediately if a timeout occurs
       req.end()
-      var err = JSON.stringify({message: 'The request timed out after ' + self.timeOut / 1000 + ' seconds.'})
-      catchErrors(task, err, uri, cb)
+      var err = {message: 'The request timed out after ' + self.timeOut / 1000 + ' seconds.'}
+      self.catchErrors(task, err, uri, cb)
     })
 
     // we need this error catch to handle ECONNRESET
     req.on('error', function (err) {
-      catchErrors(task, err, uri, cb)
+      self.catchErrors(task, err, uri, cb)
     })
-
-    req.end()
-  } catch(e) {
-    catchErrors(task, e, uri, cb)
+  } catch (e) {
+    self.catchErrors(task, e, uri, cb)
   }
+}
 
-  // Catch any errors and either retry the request or fail it
-  var catchErrors = function (task, e, url, cb) {
-    if (task.retry && task.retry === 3) {
-      try {
-        var jsonErr = JSON.parse(e)
-        this._abortPaging('Failed to request a page of features', url, jsonErr.message, jsonErr.code, cb)
-      } catch (parseErr) {
-        this._abortPaging('Failed to request a page of features', url, parseErr, null, cb)
-      }
-      return
-    }
-    // immediately kill
-    if (!task.retry) {
-      task.retry = 1
+/* Decodes a response for features
+ * @param {object} res - the response received from the GIS Server
+ * @param {array} data - an array of chunks received from the server
+ * @param {function} callback - calls back with either an error or the decoded feature json
+ */
+FeatureService.prototype._decode = function (res, data, callback) {
```

Comment:
`_decode` here does not match `decode` above
https://github.com/koopjs/featureservice/pull/9/files#diff-168726dbe96b3ce427e7fedce31bb0bcR411